### PR TITLE
Run DNS lookup in a separate thread

### LINKED ISSUE #36: The main thread gets blocked during DNS lookup when the connection is created

### PR DESCRIPTION
**Pull request by britzl**
Fixes #36

### CODE REVIEW

File: websocket/src/websocket.cpp [hunk 1]
```diff
@@ -547,11 +556,7 @@ static dmExtension::Result AppInitialize(dmExtension::AppParams* params)
     g_Websocket.m_Initialized = 1;
     if (!g_Websocket.m_Pool)
     {
-        if (!g_Websocket.m_Pool)
-        {
-            dmLogInfo("pool is null!");
-            dmConnectionPool::Delete(g_Websocket.m_Pool);
-        }
+        dmConnectionPool::Delete(g_Websocket.m_Pool);
```
Comment (JCash) [inline websocket/src/websocket.cpp]: Why do we even delete something that is null?

File: websocket/src/websocket.cpp [hunk 2]
```diff
@@ -799,20 +824,10 @@ static dmExtension::Result OnUpdate(dmExtension::Params* params)
...
-                continue;
-            }
-            conn->m_Socket = dmConnectionPool::GetSocket(g_Websocket.m_Pool, conn->m_Connection);
-            conn->m_SSLSocket = dmConnectionPool::GetSSLSocket(g_Websocket.m_Pool, conn->m_Connection);
-            SetState(conn, STATE_HANDSHAKE_WRITE);
+            conn->m_ConnectionThread = dmThread::New((dmThread::ThreadStart)ConnectionWorker, 0x80000, conn, "WebSocketConnectionThread");
 #endif
+            SetState(conn, STATE_CONNECTING);
         }
         else if (STATE_CONNECTING == conn->m_State)
```
Comment (JCash) [inline websocket/src/websocket.cpp:826]: Do we need to deal with our own timeout?
What happens if we call CLOSE_CONN, and the connection+thread is still in use?

**Review by JCash — CHANGES_REQUESTED** [review]
Looks good.
However, I think there's a case where we don't handle the shutdown of a connection.

File: websocket/src/websocket.cpp (hunk 1 above)
Comment (britzl) [reply websocket/src/websocket.cpp]: Oh, yes, I thought I had removed that completely!

Comment (britzl) [reply websocket/src/websocket.cpp]: Did cleanup and early exit if pool creation failed.

File: websocket/src/websocket.cpp (hunk 2 above)
Comment (britzl) [reply websocket/src/websocket.cpp:826]: Good point. Skipping duplicate timeout check on all platforms but emscripten

File: websocket/src/websocket.cpp [hunk 3]
```diff
@@ -520,6 +529,7 @@ static dmExtension::Result AppInitialize(dmExtension::AppParams* params)
     g_Websocket.m_Connections.SetCapacity(4);
     g_Websocket.m_Channel = 0;
     g_Websocket.m_Pool = 0;
+    g_Websocket.m_Initialized = 0;
```
Comment (JCash) [inline websocket/src/websocket.cpp:532]: 👍

File: websocket/src/websocket.cpp [hunk 4]
```diff
@@ -532,6 +542,7 @@ static dmExtension::Result AppInitialize(dmExtension::AppParams* params)
     if (dmConnectionPool::RESULT_OK != result)
     {
         dmLogError("Failed to create connection pool: %d", result);
+        return dmExtension::RESULT_INIT_ERROR;
```
Comment (JCash) [inline websocket/src/websocket.cpp:545]: 👍

File: websocket/src/websocket.cpp [hunk 5]
```diff
@@ -799,28 +818,20 @@ static dmExtension::Result OnUpdate(dmExtension::Params* params)
...
         }
         else if (STATE_CONNECTING == conn->m_State)
         {
+#if defined(__EMSCRIPTEN__)
             if (CheckConnectTimeout(conn))
             {
                 CLOSE_CONN("Connect sequence timed out");
                 continue;
             }
+#endif
```
Comment (JCash) [inline websocket/src/websocket.cpp:834]: 👍

**Review by JCash — APPROVED** [review]
Looks good

**Pull request merged**